src/main.cairo

```diff
@@ -1,32 +1,231 @@
+@storage_var
+func admin_address() -> (address: felt) {
+}
+
+@storage_var
+func naming_contract() -> (contract_address: felt) {
+}
+
+@storage_var
+func pricing_contract() -> (contract_address: felt) {
+}
+
+@storage_var
+func is_renewing(renewer: felt, domain: felt) -> (bool: felt) {
+}
 
 @storage_var
-func balance() -> (res: felt) {
+func voted_upgrade(user: felt, upgrade_id: felt, implementation_hash: felt) -> (bool: felt) {
+}
+
+@storage_var
+func blacklisted(upgrade_id: felt) -> (bool: felt) {
+}
+
+//
+// Events
+//
+
+@event
+func toggled_renewal(domain: felt, renewer: felt, value: felt) {
+}
+
+@event
+func domain_renewed(domain: felt, renewer: felt, days: felt) {
+}
+
+@event
+func voted(caller: felt, upgrade_id: felt, implementation_hash: felt, vote: felt) {
 }
 
 @external
-func increase_balance{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
-    amount: felt
+func initializer{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    admin: felt, 
+    naming_address: felt, 
+    pricing_address: felt
 ) {
-    with_attr error_message("Amount must be positive. Got: {amount}.") {
-        assert_nn(amount);
-    }
+    // Can only be called if there is no admin
+    let (current_admin) = admin_address.read();
+    assert current_admin = 0;
+
+    admin_address.write(admin);
+    naming_contract.write(naming_address);
+    pricing_contract.write(pricing_address);
 
-    let (res) = balance.read();
-    balance.write(res + amount);
+    Proxy.initializer(admin);
     return ();
 }
 
+// @notice Get the status of renewals for a domain
+// @param domain Domain to get status of
+// @param user User to get status of
 @view
-func get_balance{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() -> (res: felt) {
-    let (res) = balance.read();
+func will_renew{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    domain: felt, 
+    renewer: felt
+) -> (res: felt) {
+    let (res) = is_renewing.read(renewer, domain);
     return (res,);
 }
 
-@constructor
-func constructor{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() {
-    balance.write(0);
+// @notice Get the status of an upgrade vote
+// @param upgrade Upgrade to get status of
+// @param user User to get status of
+@view
+func has_voted_upgrade{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    user: felt,
+    upgrade_id: felt, 
+    implementation_hash: felt,
+) -> (res: felt) {
+    let (res) = voted_upgrade.read(user, upgrade_id, implementation_hash);
+    return (res,);
+}
+
+//
+// Externals
+//
+
+// @notice Activate or deactivate renewals for a domain
+// @param domain Domain to activate or deactivate renewals for
+// @dev callable by owner of domain
+@external
+func toggle_renewals{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    domain: felt,
+) {
+    alloc_locals;
+    let (caller) = get_caller_address();
+    let (contract) = naming_contract.read();
```

Comment:
You don't use the contract variable here ?

Reply:
indeed, I removed it